cpp/tests/community/k_truss_test.cu

```diff
@@ -0,0 +1,408 @@
+ public:
+  Tests_KTruss() {}
+
+  static void SetUpTestCase() {}
+  static void TearDownTestCase() {}
+
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+
+  template <typename vertex_t, typename edge_t, typename weight_t>
+  std::tuple<rmm::device_uvector<vertex_t>,
+             rmm::device_uvector<vertex_t>,
+             std::optional<rmm::device_uvector<weight_t>>>
+  k_truss_reference(
+    raft::handle_t const& handle,
+    cugraph::graph_view_t<vertex_t, edge_t, false, false> const& graph_view,
+    std::optional<cugraph::edge_property_view_t<edge_t, weight_t const*>> edge_weight_view,
+    edge_t k)
+  {
+    rmm::device_uvector<vertex_t> d_src(0, handle.get_stream());
+    rmm::device_uvector<vertex_t> d_dst(0, handle.get_stream());
+    std::optional<rmm::device_uvector<weight_t>> d_wgt{std::nullopt};
+
+    // Decompress the edgelist - For Rmat generated graph, remove self loop and multi edges
+    std::tie(d_src, d_dst, d_wgt, std::ignore) = decompress_to_edgelist(
+      handle,
+      graph_view,
+      edge_weight_view ? std::make_optional(*edge_weight_view) : std::nullopt,
+      std::optional<cugraph::edge_property_view_t<edge_t, edge_t const*>>{std::nullopt},
+      std::optional<raft::device_span<vertex_t const>>(std::nullopt));
+
+    // get host vectors
+
+    std::vector<vertex_t> h_src(d_src.size());
+    raft::update_host(h_src.data(), d_src.data(), d_src.size(), handle.get_stream());
+    std::vector<vertex_t> h_dst(d_src.size());
+    raft::update_host(h_dst.data(), d_dst.data(), d_dst.size(), handle.get_stream());
+
+    std::vector<weight_t> h_wgt(d_src.size());
+    if (edge_weight_view) {
+      raft::update_host(h_wgt.data(), (*d_wgt).data(), (*d_wgt).size(), handle.get_stream());
+    }
+
+    std::vector<vertex_t> h_src_(h_src.size());
+    std::vector<vertex_t> h_dst_(h_dst.size());
+    // FIXME: Instead extract the vertices from the graph
+    std::vector<vertex_t> dup_vertices(2 * h_src.size());
+
+    thrust::copy(thrust::host, h_src.begin(), h_src.end(), dup_vertices.begin());
+    thrust::copy(thrust::host, h_dst.begin(), h_dst.end(), dup_vertices.begin() + h_src.size());
+
+    std::set<vertex_t> vertices(dup_vertices.begin(), dup_vertices.end());
+
+    auto n_dropped = 1;
+
+    auto edge_first = thrust::make_zip_iterator(thrust::make_tuple(h_src.begin(), h_dst.begin()));
+
+    auto weighted_edge_first =
+      thrust::make_zip_iterator(thrust::make_tuple(h_src.begin(), h_dst.begin(), h_wgt.begin()));
+
+    thrust::sort(thrust::host, edge_first, edge_first + h_src.size());
+
+    while (n_dropped > 0) {
+      n_dropped = 0;
+      std::set<vertex_t> seen;
+      // Go over all the vertices
+      for (auto u = vertices.begin(); u != vertices.end(); ++u) {
+        // Find all neighbours of u
+        vertex_t idx_start   = 0;
+        vertex_t idx_end     = 0;
+        auto itr_lower_range = thrust::lower_bound(thrust::host, h_src.begin(), h_src.end(), *u);
```

Comment:
Use std::lower_bound, std::distance, and std;:upper_bound in the reference implementation.

Reply:
This is no longer needed as I am now operating on a `csr` format. Through the `offsets` array,  I can figure out the ranges.